sympy/discrete/transforms.py

```diff
@@ -0,0 +1,122 @@
+"""
+Discrete Fourier Transform, Walsh Hadamard Transform,
+Number Theoretic Transform, Zeta Transform, Mobius Transform
+"""
+from __future__ import print_function, division
+
+from sympy.core import S, Symbol, sympify
+from sympy.core.compatibility import as_int, range, iterable
+from sympy.core.function import expand, expand_mul
+from sympy.core.numbers import pi, I
+from sympy.functions.elementary.exponential import exp
+from sympy.functions.elementary.trigonometric import sin, cos
+from sympy.ntheory import isprime, primitive_root
+from sympy.utilities.iterables import ibin
+
+
+#----------------------------------------------------------------------------#
+#                                                                            #
+#                         Discrete Fourier Transform                         #
+#                                                                            #
+#----------------------------------------------------------------------------#
+
+def _fourier_transform(seq, symbolic, inverse=False):
+    """Utility function for the Discrete Fourier Transform (DFT)"""
+
+    if not iterable(seq):
+        raise TypeError("Expected a sequence of numeric coefficients " +
+                        "for Fourier Transform")
+
+    a = [sympify(arg) for arg in seq]
+    if any(x.has(Symbol) for x in a):
+        raise ValueError("Expected non-symbolic coefficients")
+
+    n = len(a)
+    if n < 2:
+        return a
+
+    b = n.bit_length() - 1
+    if n&(n - 1): # not a power of 2
+        b += 1
+        n = 2**b
+
+    a += [S.Zero]*(n - len(a))
+    for i in range(1, n):
+        j = int(ibin(i, b, str=True)[::-1], 2)
+        if i < j:
+            a[i], a[j] = a[j], a[i]
+
+    ang = -2*pi/n if inverse else 2*pi/n
+    w = None
+    if symbolic:
+        w = [cos(ang*i) + I*sin(ang*i) for i in range(n // 2)]
+    else:
+        w = [exp(I*ang*i).evalf() for i in range(n // 2)]
```

Comment:
As I noted on gitter, I think this will already just do the right thing if the input is evalf'd first. Is this not the case?

Reply:
I guess it doesn't for exp, but it does for sin and cos. Maybe `exp` should be fixed here. Or you could just detect if the input is floating point (actually `Float + Float*I`) automatically.

Reply:
> As I noted on gitter, I think this will already just do the right thing if the input is evalf'd first. Is this not the case?

Yes, apparently it does. So is it suggested to convert argument to `Float` explicity (to handle the integer case, where it remains unevaluated).

> Maybe exp should be fixed here

Does this mean that its behaviour should be similar to `sin` and `cos` for floating point inputs?
(Currently both `sin`/`cos` yield a float if input is `floating` (`real`/`complex`), but this is not the case with `exp` which as you mentioned evaluates only for `real floats`)

Reply:
Is it necessary to use `exp` at all if `sin` and `cos` can be converted to Floats (by making `ang` a Float of suitable precision)?

Reply:
Fair point. In that case, we might just need to devise a way for deciding a suitable precision.

Reply:
It's probably more efficient to use exp. But also not really worth separate code paths IMO, since it should just do the right thing on floating point inputs. 

> Fair point. In that case, we might just need to devise a way for deciding a suitable precision.

I think if you let exp evaluate itself (fix the issue I noted on gitter), then the precision won't be an issue. Because the exp will just evaluate to whatever precision the input is, just like `sin` and `cos` do now.